src/js/parsers/section.js

```diff
@@ -155,29 +140,44 @@ class SectionParser {
     }
 
     // handle closing the current section and starting a new one if we hit a
-    // new-section-creating element
-    if (this.state.section && !this.state.section.isListItem && !isTextNode(node) && node.tagName) {
-      // handle lists nested inside wrappers
-      if (this.state.section.isListSection) {
-        this.parseListItems(node.childNodes);
-        return;
-      }
-
+    // new-section-creating element.
+    if (this.state.section && !isTextNode(node) && node.tagName) {
       let tagName = normalizeTagName(node.tagName);
-      if (contains(VALID_MARKUP_SECTION_TAGNAMES, tagName) || contains(VALID_LIST_SECTION_TAGNAMES, tagName)) {
-        // avoid creating empty paragraphs due to wrappers elements around
+      let isNestedListSection = contains(VALID_LIST_SECTION_TAGNAMES, tagName)
+        && this.state.section.isListItem;
+
+      if (
+        contains(VALID_MARKUP_SECTION_TAGNAMES, tagName) ||
+        (contains(VALID_LIST_SECTION_TAGNAMES, tagName) && !isNestedListSection) ||
+        contains(VALID_LIST_ITEM_TAGNAMES, tagName)
+      ) {
+        // don't break out of the list for list items that contain a single <p>.
+        // deals with typical case of <li><p>Text</p></li><li><p>Text</p></li>
```

Comment:
I added this because I've seen other editors and markdown parsers create this type of structure so I figured it would be good to add specific handling. Otherwise the new behaviour would end up dropping the list altogether, instead creating paragraph sections in place of list sections.

Reply:
@kevinansfield As far as I recall the section parser should only be parsing DOM we create or introduced by `contexteditable` interactions. For example paste events and importing HTML are handled via the DOM parser.

So the implication here is that browsers may insert `p` tags inside lists as the user interacts. Is that what you've seen?

Reply:
The DOM parser only grabs top-level elements and relies on the section parser to deal with all the nested elements, in this case the `LI` and `P` are always nested elements as they're inside a `UL`. The changes are primarily to do with handling pastes and using the parsers externally for html->mobiledoc.

Reply:
Yeah, as implemented here:

https://github.com/bustle/mobiledoc-kit/blob/28a746d694e08c7cfc426a4e71a8667b845a1a4d/src/js/parsers/dom.js#L140

You are correct. I feel like I should have caught this earlier, but if these are explicitly corrections to the paste parsing then we should implement them as corrections in the DOM parser before it passes content to the section parser. The section parser (in theory) should only be concerned with contexteditable-type inputs, not with any plausible HTML input from a paste. The latter is basically an unbounded set of concerns, and it is nice to keep them out of the editor parser.